File: backend/src/main/java/com/healthcare/app/service/AuthService.java
package com.healthcare.app.service;

import com.healthcare.app.config.JwtUtil;
import com.healthcare.app.entity.*;
import com.healthcare.app.repository.*;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.Optional;
import java.util.Random;
import java.util.UUID;

@Service
public class AuthService {

    private final PatientRepository patientRepository;
    private final AdminRepository adminRepository;
    private final HospitalRepository hospitalRepository;
    private final DoctorRepository doctorRepository;
    private final HospitalRegistrationRequestRepository hospitalRequestRepository;
    private final DoctorRegistrationRequestRepository doctorRequestRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtUtil jwtUtil;
    private final OtpRepository otpRepository;
    private final EmailService emailService;

    public AuthService(PatientRepository patientRepository,
                       AdminRepository adminRepository,
                       HospitalRepository hospitalRepository,
                       DoctorRepository doctorRepository,
                       HospitalRegistrationRequestRepository hospitalRequestRepository,
                       DoctorRegistrationRequestRepository doctorRequestRepository,
                       PasswordEncoder passwordEncoder,
                       JwtUtil jwtUtil,
                       OtpRepository otpRepository,
                       EmailService emailService) {
        this.patientRepository = patientRepository;
        this.adminRepository = adminRepository;
        this.hospitalRepository = hospitalRepository;
        this.doctorRepository = doctorRepository;
        this.hospitalRequestRepository = hospitalRequestRepository;
        this.doctorRequestRepository = doctorRequestRepository;
        this.passwordEncoder = passwordEncoder;
        this.jwtUtil = jwtUtil;
        this.otpRepository = otpRepository;
        this.emailService = emailService;
    }

    public Object register(RegistrationRequest request) {
        String role = request.getRole().toUpperCase();

        switch (role) {
            case "USER":
                Patient patient = new Patient();
                patient.setName(request.getName());
                patient.setEmail(request.getEmail());
                patient.setPassword(request.getPassword());
                patient.setPhone(request.getPhone());
                patient.setDateOfBirth(request.getDateOfBirth());
                patient.setGender(request.getGender());
                patient.setWeight(Double.parseDouble(request.getWeight()));
                return registerPatient(patient);
            case "DOCTOR":
                DoctorRegistrationRequest doctorRequest = new DoctorRegistrationRequest();
                doctorRequest.setName(request.getName());
                doctorRequest.setEmail(request.getEmail());
                doctorRequest.setPassword(request.getPassword());
                doctorRequest.setPhone(request.getPhone());
                doctorRequest.setSpecialization(request.getSpecialization());
                doctorRequest.setLicenseNumber(request.getLicenseNumber());
                return registerDoctor(doctorRequest);
            case "HOSPITAL":
                HospitalRegistrationRequest hospitalRequest = new HospitalRegistrationRequest();
                hospitalRequest.setName(request.getName());
                hospitalRequest.setEmail(request.getEmail());
                hospitalRequest.setPassword(request.getPassword());
                hospitalRequest.setPhone(request.getPhone());
                hospitalRequest.setHospitalName(request.getHospitalName());
                hospitalRequest.setAddress(request.getAddress());
                hospitalRequest.setCity(request.getCity());
                hospitalRequest.setState(request.getState());
                hospitalRequest.setZipCode(request.getZipCode());
                return registerHospital(hospitalRequest);
            default:
                throw new RuntimeException("Invalid role");
        }
    }

    public HospitalRegistrationRequest registerHospital(HospitalRegistrationRequest request) {
        if (hospitalRequestRepository.findByEmail(request.getEmail()).isPresent()) {
            throw new RuntimeException("Email already registered");
        }
        request.setPassword(passwordEncoder.encode(request.getPassword()));
        return hospitalRequestRepository.save(request);
    }

    public DoctorRegistrationRequest registerDoctor(DoctorRegistrationRequest request) {
        if (doctorRequestRepository.findByEmail(request.getEmail()).isPresent()) {
            throw new RuntimeException("Email already registered");
        }
        request.setPassword(passwordEncoder.encode(request.getPassword()));
        return doctorRequestRepository.save(request);
    }

    public Patient registerPatient(Patient patient) {
        if (patientRepository.findByEmail(patient.getEmail()).isPresent()) {
            throw new RuntimeException("Email already registered");
        }
        patient.setRole("USER");
        patient.setPid(generateNextPatientId());
        patient.setPatientId(UUID.randomUUID().toString());
        patient.setPassword(passwordEncoder.encode(patient.getPassword()));
        return patientRepository.save(patient);
    }

    private String generateNextPatientId() {
        Long maxId = patientRepository.findMaxId();
        if (maxId == null) {
            return "P001";
        }
        return String.format("P%03d", maxId + 1);
    }

    private String generateNextDoctorId() {
        Long maxId = doctorRepository.findMaxId();
        if (maxId == null) {
            return "D001";
        }
        return String.format("D%03d", maxId + 1);
    }

    private String generateNextHospitalId() {
        Long maxId = hospitalRepository.findMaxId();
        if (maxId == null) {
            return "H001";
        }
        return String.format("H%03d", maxId + 1);
    }

    public String login(String email, String password) {
        // Check in Admins
        var adminOpt = adminRepository.findByEmail(email);
        if (adminOpt.isPresent()) {
            Admin admin = adminOpt.get();
            if (passwordEncoder.matches(password, admin.getPassword())) {
                return jwtUtil.generateToken(admin.getId().toString(), admin.getRole());
            }
        }

        // Check in Patients
        var patientOpt = patientRepository.findByEmail(email);
        if (patientOpt.isPresent()) {
            Patient patient = patientOpt.get();
            if (passwordEncoder.matches(password, patient.getPassword())) {
                return jwtUtil.generateToken(patient.getPid(), patient.getRole());
            }
        }

        // Check in Hospitals
        var hospitalOpt = hospitalRepository.findByEmail(email);
        if (hospitalOpt.isPresent()) {
            Hospital hospital = hospitalOpt.get();
            if (passwordEncoder.matches(password, hospital.getPassword())) {
                return jwtUtil.generateToken(hospital.getHid(), hospital.getRole());
            }
        }

        // Check in Doctors
        var doctorOpt = doctorRepository.findByEmail(email);
        if (doctorOpt.isPresent()) {
            Doctor doctor = doctorOpt.get();
            if (passwordEncoder.matches(password, doctor.getPassword())) {
                return jwtUtil.generateToken(doctor.getDid(), doctor.getRole());
            }
        }

        throw new RuntimeException("Invalid credentials");
    }

    public void changePassword(String email, String oldPassword, String newPassword) {
        // Check in Admins
        var adminOpt = adminRepository.findByEmail(email);
        if (adminOpt.isPresent()) {
            Admin admin = adminOpt.get();
            if (passwordEncoder.matches(oldPassword, admin.getPassword())) {
                admin.setPassword(passwordEncoder.encode(newPassword));
                adminRepository.save(admin);
                return;
            } else {
                throw new RuntimeException("Invalid old password");
            }
        }

        // Check in Patients
        var patientOpt = patientRepository.findByEmail(email);
        if (patientOpt.isPresent()) {
            Patient patient = patientOpt.get();
            if (passwordEncoder.matches(oldPassword, patient.getPassword())) {
                patient.setPassword(passwordEncoder.encode(newPassword));
                patientRepository.save(patient);
                return;
            } else {
                throw new RuntimeException("Invalid old password");
            }
        }

        // Check in Hospitals
        var hospitalOpt = hospitalRepository.findByEmail(email);
        if (hospitalOpt.isPresent()) {
            Hospital hospital = hospitalOpt.get();
            if (passwordEncoder.matches(oldPassword, hospital.getPassword())) {
                hospital.setPassword(passwordEncoder.encode(newPassword));
                hospitalRepository.save(hospital);
                return;
            } else {
                throw new RuntimeException("Invalid old password");
            }
        }

        // Check in Doctors
        var doctorOpt = doctorRepository.findByEmail(email);
        if (doctorOpt.isPresent()) {
            Doctor doctor = doctorOpt.get();
            if (passwordEncoder.matches(oldPassword, doctor.getPassword())) {
                doctor.setPassword(passwordEncoder.encode(newPassword));
                doctorRepository.save(doctor);
                return;
            } else {
                throw new RuntimeException("Invalid old password");
            }
        }

        throw new RuntimeException("User not found");
    }

    @Transactional
    public void forgotPassword(String email) {
        // Check if user exists
        boolean userExists = adminRepository.findByEmail(email).isPresent() ||
                           patientRepository.findByEmail(email).isPresent() ||
                           hospitalRepository.findByEmail(email).isPresent() ||
                           doctorRepository.findByEmail(email).isPresent();

        if (!userExists) {
            throw new RuntimeException("User not found");
        }

        // Clean up expired OTPs
        otpRepository.deleteExpiredOtps(LocalDateTime.now());

        // Generate OTP
        String otp = generateOtp();

        // Save OTP
        Otp otpEntity = new Otp();
        otpEntity.setEmail(email);
        otpEntity.setOtpCode(otp);
        otpEntity.setCreatedAt(LocalDateTime.now());
        otpEntity.setExpiresAt(LocalDateTime.now().plusMinutes(10));
        otpRepository.save(otpEntity);

        // Send email
        emailService.sendOtpEmail(email, otp);
    }

    @Transactional
    public void resetPassword(String email, String otp, String newPassword) {
        // Verify OTP
        Optional<Otp> otpEntity = otpRepository.findByEmailAndOtpCodeAndUsedFalse(email, otp);
        if (otpEntity.isEmpty()) {
            throw new RuntimeException("Invalid or expired OTP");
        }

        Otp validOtp = otpEntity.get();
        if (validOtp.getExpiresAt().isBefore(LocalDateTime.now())) {
            throw new RuntimeException("OTP has expired");
        }

        // Mark OTP as used
        otpRepository.markOtpAsUsed(email, otp);

        // Update password for the user
        updateUserPassword(email, newPassword);
    }

    private String generateOtp() {
        Random random = new Random();
        return String.format("%06d", random.nextInt(999999));
    }

    private void updateUserPassword(String email, String newPassword) {
        // Check in Admins
        var adminOpt = adminRepository.findByEmail(email);
        if (adminOpt.isPresent()) {
            Admin admin = adminOpt.get();
            admin.setPassword(passwordEncoder.encode(newPassword));
            adminRepository.save(admin);
            return;
        }

        // Check in Patients
        var patientOpt = patientRepository.findByEmail(email);
        if (patientOpt.isPresent()) {
            Patient patient = patientOpt.get();
            patient.setPassword(passwordEncoder.encode(newPassword));
            patientRepository.save(patient);
            return;
        }

        // Check in Hospitals
        var hospitalOpt = hospitalRepository.findByEmail(email);
        if (hospitalOpt.isPresent()) {
            Hospital hospital = hospitalOpt.get();
            hospital.setPassword(passwordEncoder.encode(newPassword));
            hospitalRepository.save(hospital);
            return;
        }

        // Check in Doctors
        var doctorOpt = doctorRepository.findByEmail(email);
        if (doctorOpt.isPresent()) {
            Doctor doctor = doctorOpt.get();
            doctor.setPassword(passwordEncoder.encode(newPassword));
            doctorRepository.save(doctor);
            return;
        }

        throw new RuntimeException("User not found");
    }
}

File: frontend/src/services/AuthService.ts
import { Patient, Doctor, Hospital, Admin, HospitalRegistrationRequest, DoctorRegistrationRequest, Otp } from '../types';
import { MedicalRecord, Notification, Reminder, Review, Availability } from '../types/backend-types';
import {
  patientRepository,
  doctorRepository,
  hospitalRepository,
  adminRepository,
  hospitalRequestRepository,
  doctorRequestRepository,
  otpRepository
} from '../lib/data-store';

interface RegistrationRequest {
  role: string;
  name: string;
  email: string;
  password: string;
  phone?: string;
  dateOfBirth?: string;
  gender?: string;
  weight?: string;
  specialization?: string;
  licenseNumber?: string;
}

interface ChangePasswordRequest {
  oldPassword: string;
  newPassword: string;
}

export class AuthService {

  // Simple password encoding (in real app, use proper hashing)
  private encodePassword(password: string): string {
    return btoa(password); // Base64 encoding for demo
  }

  private decodePassword(encodedPassword: string): string {
    return atob(encodedPassword);
  }

  private verifyPassword(rawPassword: string, encodedPassword: string): boolean {
    return this.decodePassword(encodedPassword) === rawPassword;
  }

  register(request: RegistrationRequest): any {
    const role = request.role.toUpperCase();

    switch (role) {
      case "USER":
        return this.registerPatient({
          id: this.generatePatientId(),
          pid: this.generatePatientId(),
          patientId: this.generatePatientId(),
          name: request.name,
          email: request.email,
          password: this.encodePassword(request.password),
          phone: request.phone || '',
          age: 0,
          gender: request.gender as 'Male' | 'Female' | 'Other' || 'Male',
          weight: parseFloat(request.weight || '0'),
          dateOfBirth: request.dateOfBirth || '',
          role: 'USER',
          createdAt: new Date().toISOString(),
        });
      case "DOCTOR":
        return this.registerDoctor({
          id: this.generateDoctorId(),
          name: request.name,
          email: request.email,
          password: this.encodePassword(request.password),
          phone: request.phone || '',
          specialization: request.specialization || '',
          licenseNumber: request.licenseNumber || '',
          role: 'DOCTOR',
          status: 'pending',
          experience: 0,
          rating: 0,
          available: true,
          consultationFee: 0,
          createdAt: new Date().toISOString(),
        });
      case "HOSPITAL":
        return this.registerHospital({
          id: this.generateHospitalId(),
          name: request.name,
          hospitalName: request.name,
          email: request.email,
          password: this.encodePassword(request.password),
          phone: request.phone || '',
          address: '',
          city: '',
          state: '',
          zipCode: '',
          role: 'HOSPITAL',
          status: 'pending',
          totalDoctors: 0,
          facilities: [],
          registrationNumber: '',
          createdAt: new Date().toISOString(),
        });
      default:
        throw new Error("Invalid role");
    }
  }

  registerPatient(patient: Patient): Patient {
    if (patientRepository.findByEmail(patient.email)) {
      throw new Error("Email already registered");
    }
    return patientRepository.save(patient);
  }

  registerHospital(request: HospitalRegistrationRequest): HospitalRegistrationRequest {
    if (hospitalRequestRepository.findByEmail(request.email)) {
      throw new Error("Email already registered");
    }
    return hospitalRequestRepository.save(request);
  }

  registerDoctor(request: DoctorRegistrationRequest): DoctorRegistrationRequest {
    if (doctorRequestRepository.findByEmail(request.email)) {
      throw new Error("Email already registered");
    }
    return doctorRequestRepository.save(request);
  }

  login(email: string, password: string): string {
    // Check in Admins
    const admin = adminRepository.findByEmail(email);
    if (admin && this.verifyPassword(password, admin.password)) {
      return this.generateToken(admin.email, admin.role);
    }

    // Check in Patients
    const patient = patientRepository.findByEmail(email);
    if (patient && patient.password && this.verifyPassword(password, patient.password)) {
      return this.generateToken(patient.email, patient.role || 'USER');
    }

    // Check in Hospitals
    const hospital = hospitalRepository.findByEmail(email);
    if (hospital && hospital.password && this.verifyPassword(password, hospital.password)) {
      return this.generateToken(hospital.email, hospital.role || 'HOSPITAL');
    }

    // Check in Doctors
    const doctor = doctorRepository.findByEmail(email);
    if (doctor && doctor.password && this.verifyPassword(password, doctor.password)) {
      return this.generateToken(doctor.email, doctor.role || 'DOCTOR');
    }

    throw new Error("Invalid credentials");
  }

  changePassword(email: string, oldPassword: string, newPassword: string): void {
    // Check in Admins
    let admin = adminRepository.findByEmail(email);
    if (admin && this.verifyPassword(oldPassword, admin.password)) {
      admin.password = this.encodePassword(newPassword);
      adminRepository.save(admin);
      return;
    }

    // Check in Patients
    let patient = patientRepository.findByEmail(email);
    if (patient && patient.password && this.verifyPassword(oldPassword, patient.password)) {
      patient.password = this.encodePassword(newPassword);
      patientRepository.save(patient);
      return;
    }

    // Check in Hospitals
    let hospital = hospitalRepository.findByEmail(email);
    if (hospital && hospital.password && this.verifyPassword(oldPassword, hospital.password)) {
      hospital.password = this.encodePassword(newPassword);
      hospitalRepository.save(hospital);
      return;
    }

    // Check in Doctors
    let doctor = doctorRepository.findByEmail(email);
    if (doctor && doctor.password && this.verifyPassword(oldPassword, doctor.password)) {
      doctor.password = this.encodePassword(newPassword);
      doctorRepository.save(doctor);
      return;
    }

    throw new Error("User not found or invalid password");
  }

  forgotPassword(email: string): void {
    // Check if user exists
    const userExists = adminRepository.findByEmail(email) ||
                      patientRepository.findByEmail(email) ||
                      hospitalRepository.findByEmail(email) ||
                      doctorRepository.findByEmail(email);

    if (!userExists) {
      throw new Error("User not found");
    }

    const otp = this.generateOtp();
    const otpEntity: Otp = {
      id: 0, // Will be set by repository
      email,
      otpCode: otp,
      expiresAt: new Date(Date.now() + 10 * 60 * 1000), // 10 minutes
      used: false,
    };

    otpRepository.save(otpEntity);

    // In real app, send email
    console.log(`OTP for ${email}: ${otp}`);
  }

  resetPassword(email: string, otp: string, newPassword: string): void {
    const otpEntity = otpRepository.findByEmailAndOtpCodeAndUsedFalse(email, otp);
    if (!otpEntity) {
      throw new Error("Invalid or expired OTP");
    }

    if (otpEntity.expiresAt < new Date()) {
      throw new Error("OTP has expired");
    }

    // Mark OTP as used
    otpRepository.markOtpAsUsed(email, otp);

    // Update password
    this.updateUserPassword(email, newPassword);
  }

  private updateUserPassword(email: string, newPassword: string): void {
    const encodedPassword = this.encodePassword(newPassword);

    // Check in Admins
    let admin = adminRepository.findByEmail(email);
    if (admin) {
      admin.password = encodedPassword;
      adminRepository.save(admin);
      return;
    }

    // Check in Patients
    let patient = patientRepository.findByEmail(email);
    if (patient) {
      patient.password = encodedPassword;
      patientRepository.save(patient);
      return;
    }

    // Check in Hospitals
    let hospital = hospitalRepository.findByEmail(email);
    if (hospital) {
      hospital.password = encodedPassword;
      hospitalRepository.save(hospital);
      return;
    }

    // Check in Doctors
    let doctor = doctorRepository.findByEmail(email);
    if (doctor) {
      doctor.password = encodedPassword;
      doctorRepository.save(doctor);
      return;
    }

    throw new Error("User not found");
  }

  getCurrentUser(email: string): any {
    // Check in Admins
    const admin = adminRepository.findByEmail(email);
    if (admin) {
      return {
        id: admin.id.toString(),
        name: admin.name,
        email: admin.email,
        role: admin.role,
      };
    }

    // Check in Patients
    const patient = patientRepository.findByEmail(email);
    if (patient) {
      return {
        id: patient.pid,
        name: patient.name,
        email: patient.email,
        role: patient.role,
      };
    }

    // Check in Hospitals
    const hospital = hospitalRepository.findByEmail(email);
    if (hospital) {
      return {
        id: hospital.hid,
        name: hospital.name,
        email: hospital.email,
        role: hospital.role,
      };
    }

    // Check in Doctors
    const doctor = doctorRepository.findByEmail(email);
    if (doctor) {
      return {
        id: doctor.did,
        name: doctor.name,
        email: doctor.email,
        role: doctor.role,
      };
    }

    return null;
  }

  private generateToken(email: string, role: string): string {
    // Simple token generation for demo (in real app, use JWT)
    const payload = { email, role, exp: Date.now() + 24 * 60 * 60 * 1000 }; // 24 hours
    return btoa(JSON.stringify(payload));
  }

  private generateOtp(): string {
    return Math.floor(100000 + Math.random() * 900000).toString();
  }

  private generatePatientId(): string {
    const maxId = patientRepository.findMaxId();
    return `P${(maxId + 1).toString().padStart(3, '0')}`;
  }

  private generateDoctorId(): string {
    const maxId = doctorRepository.findMaxId();
    return `D${(maxId + 1).toString().padStart(3, '0')}`;
  }

  private generateHospitalId(): string {
    const maxId = hospitalRepository.findMaxId();
    return `H${(maxId + 1).toString().padStart(3, '0')}`;
  }
}
